src/Functions/date_trunc.cpp

```diff
@@ -0,0 +1,167 @@
+        };
+
+        if (arguments.size() == 2)
+        {
+            check_first_argument();
+            check_second_argument();            
+        }
+        else if (arguments.size() == 3)
+        {
+            check_first_argument();
+            check_second_argument();            
+            check_timezone_argument();
+        }
+        else
+        {
+            throw Exception(
+                "Number of arguments for function " + getName() + " doesn't match: passed " + toString(arguments.size())
+                    + ", should be 2 or 3",
+                ErrorCodes::NUMBER_OF_ARGUMENTS_DOESNT_MATCH);
+        }
+
+        if (result_type_is_date)
+            return std::make_shared<DataTypeDate>();
+        else
+            return std::make_shared<DataTypeDateTime>(extractTimeZoneNameFromFunctionArguments(arguments, 2, 1));
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+    ColumnNumbers getArgumentsThatAreAlwaysConstant() const override { return {0, 2}; }
+
+    void executeImpl(Block & block, const ColumnNumbers & arguments, size_t result, size_t input_rows_count) const override
+    {
+        Block temp_block = block;
+
+        const UInt16 interval_value = 1;
+        const ColumnPtr interval_column = ColumnConst::create(ColumnInt64::create(1, interval_value), input_rows_count);
+
+        const size_t interval_pos = temp_block.columns();
+        temp_block.insert({interval_column, std::make_shared<DataTypeInterval>(datepart_kind), ""});
+
+        if (arguments.size() == 2)
+        {
+            auto to_start_of_interval = FunctionFactory::instance().get("toStartOfInterval", context)->build(
+                {temp_block.getByPosition(arguments[1]), temp_block.getByPosition(interval_pos)});
+
+            to_start_of_interval->execute(temp_block, {arguments[1], interval_pos}, result, input_rows_count);
+        } 
+        else 
+        {
+            auto to_start_of_interval = FunctionFactory::instance().get("toStartOfInterval", context)->build(
+                {temp_block.getByPosition(arguments[1]), temp_block.getByPosition(interval_pos), 
+                    temp_block.getByPosition(arguments[2])});
+
+            to_start_of_interval->execute(temp_block, {arguments[1], interval_pos, arguments[2]}, result, input_rows_count);
+        }
+
+        block.getByPosition(result).column = std::move(temp_block.getByPosition(result).column);
+    }
+
+    bool hasInformationAboutMonotonicity() const override
+    {
+        return true;
+    }
+
+    Monotonicity getMonotonicityForRange(const IDataType &, const Field &, const Field &) const override
+    {
+        return { true, true, true };
+    }
+
+private:
+    const Context & context;    
+    mutable IntervalKind::Kind datepart_kind = IntervalKind::Kind::Second;
+};
+
+
+void registerFunctionDateTrunc(FunctionFactory & factory)
+{
+    factory.registerFunction<FunctionDateTrunc>();
```

Comment:
Let's make it case insensitive - for compatibility.

Reply:
sorry, I didn't understand what should be case insensitive - *datepart* - parameter (to accept any unit like 'hour', 'HOUR', 'Hour', 'houR',  etc) ?

Reply:
Or do you mean the function naming as suggested @hczhcz below?

Reply:
I mean `DATE_TRUNC` is introduced for compatibility with other DBMS, so the function name should be case insensitive:
`DATE_TRUNC` / `date_trunc`.

It's not necessary to introduce `dateTrunc` synonim if it's equivalent to `toStartOfInterval`, but it's Ok to have.

Reply:
date_trunc is different from toStartOfInterval (order and types of arguments) so the name `dateTrunc` and case-insensitive alias `date_trunc` is a good idea.

Reply:
@alexey-milovidov done.